### ISSUE TITLE
translated volume is cut off in the rendered image

### ISSUE THREAD

**Issue by lschmidtke** [facebookresearch/pytorch3d#786]
Hi,

I have a question regarding rendering a volume after applying a transformation via `torch.nn.functional.affine_grid` and `grid_sample`.  As a test, I have created a 3d gaussian ellipsoid and translated it along the y-axis, but there seems to be an issue on the image boundary in the rendered image, as the lower part of the volume appears to be cut off. I have instead tried to translate via the `volume_translation` argument supplied to the `volumes` constructor and it works as expected, but I'd rather implement the straightforward solution. Any help would be greatly appreciated!
Below is a minimal example that should illustrate my problem:
```
import torch
import numpy as np
from scipy.stats import multivariate_normal
import matplotlib.pyplot as plt

# Data structures and functions for rendering
from pytorch3d.structures import Volumes
from pytorch3d.renderer import (
    FoVPerspectiveCameras, 
    VolumeRenderer,
    NDCGridRaysampler,
    AbsorptionOnlyRaymarcher
)

if torch.cuda.is_available():
    device = torch.device("cuda:0")
    torch.cuda.set_device(device)
else:
    device = torch.device("cpu")

# create gaussian ellipsoid with size 32, 32, 32
x, y, z = np.mgrid[-1:1.0:32j, -1.0:1.0:32j, -1.0:1.0:32j]
xy = np.column_stack([x.flat, y.flat, z.flat])

mu = np.array([0.0, 0.0, 0.0])
sigma = np.array([.1, .3, .2])

covariance = np.diag(sigma**2)

z = multivariate_normal.pdf(xy, mean=mu, cov=covariance)

# Reshape back to a (30, 30) grid.
z = z.reshape(x.shape)
# normalize
z = z/z.max()

# define volume renderer
raysampler = NDCGridRaysampler(
    image_width=32,
    image_height=32,
    n_pts_per_ray=50,
    min_depth=0.1,
    max_depth=6,
)
raymarcher = AbsorptionOnlyRaymarcher()

renderer = VolumeRenderer(
    raysampler=raysampler, raymarcher=raymarcher,
).to(device)

density = torch.from_numpy(z[None, None, :, :]).to(device).float()
features = torch.ones(1, 3, 32, 32, 32).to(device).float()

#define translation and construct affine matrix
tx = 0
ty = 0.7
tz = 0
M = torch.tensor([[1, 0, 0, tx],
                 [0, 1, 0, ty],
                 [0, 0, 1, tz]]).to(device)
M = M.view(1, 3, 4)

# transform via torch affine_grid and grid_sample
grid = torch.nn.functional.affine_grid(M, size=torch.Size((1, 1, 32, 32, 32)))
transformed_density = torch.nn.functional.grid_sample(density, grid)

# define volume and camera
volumes = Volumes(
            densities = transformed_density,
            features = features,
            voxel_size=0.1,
        )

cam_t = torch.tensor([[0, 0, 4]]).float().to(device)
cameras = FoVPerspectiveCameras(T=cam_t, device=device)

# plot rendered image, cut-off 
rendered = renderer(cameras, volumes)
plt.figure()
img = rendered[0][0, :, :, 0:3].cpu().numpy()
plt.imshow(img) 
# plot volume for comparison
plt.figure()
vol_img = transformed_density[0, 0, 16].cpu().numpy()
plt.imshow(vol_img) 
```
rendered image with cutoff: (y should be inversed, but that's easily adressed))
![rendered](https://user-images.githubusercontent.com/50370594/128232895-27c12a52-29a2-4aaa-84a5-afc651e7f8eb.png)

how it's supposed to look like (slice trough  mid z (16) of the volume) 
![vol](https://user-images.githubusercontent.com/50370594/128232707-abea05a5-36d3-4d24-872c-8310ba208f2e.png)

**Comment by nikhilaravi** [comment]
What is `ro` in `transformed_density = torch.nn.functional.grid_sample(ro, grid)`?

It might be due to the coordinate system conventions for grid sample being different to PyTorch3D. 
@davnov134 do you have any suggestions

**Comment by lschmidtke** [comment]
`ro` is density, I edited the example. Yes the coordinate systems are different, hence the different directions of the translation, but I don't understand the cut-off. I also played around with the voxel size and a different camera position along translation direction, but the problem remains.

**Comment by davnov134** [comment]
If I understand correctly, the image above is an output of rendering with a perspective camera. The bottom image contains a simple orthographic projection of the volume. 

Both images are expected to be different as perspective projection will produce different image from the orthographic one, especially when the shape is quite deep and close to the camera.

Can you try to render the image with and instance of `OrthographicCameras` instead of `FOVPerspectiveCameras`?

Furthermore, the affine grid of PyTorch can have different coordinate conventions to `Volumes`, which probably results in the z-flip.

**Comment by lschmidtke** [comment]
> If I understand correctly, the image above is an output of rendering with a perspective camera. The bottom image contains a simple orthographic projection of the volume.
> 
> Both images are expected to be different as perspective projection will produce different image from the orthographic one, especially when the shape is quite deep and close to the camera.
> 
> Can you try to render the image with and instance of `OrthographicCameras` instead of `FOVPerspectiveCameras`?
> 
> Furthermore, the affine grid of PyTorch can have different coordinate conventions to `Volumes`, which probably results in the y-flip.

That is correct. The inverted y-axis is no issue, that boils down to convention. I'm concerned with the cutoff that seems to happen even though the volume is still completely in the FOV. The rendered image with `OrthographicCameras` looks fine:
![ortho](https://user-images.githubusercontent.com/50370594/129766817-4a2b9196-ef66-4271-bdfd-16938806ec1a.png)

For comparison, the rendered image with `FOVPerspectiveCameras`. You can notice that there is a weird boundary where intensity drops off:
![rendered](https://user-images.githubusercontent.com/50370594/129766977-b3d4f45f-370f-459d-b91e-9842e39cdd61.png)

**Comment by davnov134** [comment]
Thanks, in my view, the orthographic projection result is pretty much in-line with the perspective one.

First, we can see that the Ortho projection also generates an image with almost half of the opacity map outside the image boundary.

In case of the perspective projection, we are observing the volume from a higher vantage point, hence the probability density gets lost in the boundary of the voxel grid. This boundary, in case of persp. projection, is fully visible due to the perspective effect. The AO raymarcher's rays terminate at this boundary, which corresponds to the weird cut-off you are observing.

**Comment by lschmidtke** [comment]
> Thanks, in my view, the orthographic projection result is pretty much in-line with the perspective one.
> 
> First, we can see that the Ortho projection also generates an image with almost half of the opacity map outside the image boundary.
> 
> In case of the perspective projection, we are observing the volume from a higher vantage point, hence the probability density gets lost in the boundary of the voxel grid. This boundary, in case of persp. projection, is fully visible due to the perspective effect. The AO raymarcher's rays terminate at this boundary, which corresponds to the weird cut-off you are observing.

Okay, thank you, I believe that makes sense, since the rays are not perpendicular to the camera direction anymore. However, I'm still a bit confused why this does not occur when I pass `volume_translation=[0., 0.7, 0.]` in the `Volumes` constructor instead of using `affine_grid` and `grid_sample`:

```
volumes = Volumes(
            densities = transformed_density,
            features = features,
            voxel_size=0.1,
            volume_translation = [0., 0.7, 0.]
        )
```
![translated](https://user-images.githubusercontent.com/50370594/129910460-3528d79a-9bd0-41c5-a50d-37b796c948dc.png)

**Comment by davnov134** [comment]
In case of the affine_grid solution, the density is first evaluated on the non-translated grid, which causes the cut-off and then translated using the sampler. In case of volume_translation, the density is evaluated on an already translated grid, so the whole density is contained in the volume.

**Comment by lschmidtke** [comment]
> In case of the affine_grid solution, the density is first evaluated on the non-translated grid, which causes the cut-off, and then translated using the sampler.
> 
> In case of volume_translation, the density is evaluated on an already translated grid, so the whole density is contained in the volume.

thank you! closing now
